keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -1462,225 +1278,309 @@ prevented. Deleting the entire cluster also leaves resources allocated outside
-}
-
-// StructuredResourceHandle is the in-tree representation of the allocation result.
-type StructuredResourceHandle struct {
-    // VendorClassParameters are the per-claim configuration parameters
-    // from the resource class at the time that the claim was allocated.
-    VendorClassParameters runtime.Object
+    // Config contains all the configuration pieces that apply to the request
+    // and that were meant for the driver which handles these devices.
+    // They get collected during the allocation and stored here
+    // to ensure that they remain available while the claim is allocated.
+    //
+    // Entries are list in the same order as in class.config and claim.config,
+    // with class.config entries first.
+    //
+    // +optional
+    Config []DeviceConfiguration
 
-    // VendorClaimParameters are the per-claim configuration parameters
-    // from the resource claim parameters at the time that the claim was
-    // allocated.
-    VendorClaimParameters runtime.Object
+    // RequestName identifies the request in the claim which caused this
+    // device to be allocated. If a request was satisfied by multiple
+    // devices, they are indexed starting with zero in the order
+    // in which they are listed in the results.
+    RequestName string
 
-    // NodeName is the name of the node providing the necessary resources
-    // if the resources are local to a node.
-    NodeName string
+    // This name together with the driver name and the device name field
+    // identify which device was allocated.
+    //
+    // For a driver with node-local devices, it is the name of the node and
+    // the overall ID is `<driver name>/<node name>/<device name>`.
+    //
+    // For other drivers, the overall ID is `<driver name>[/pool
+    // name>]/<device name>`.
+    //
+    // Must not be longer than 253 characters and may contain one or more
+    // DNS sub-domains separated by slashes.
+    //
+    // +optional
+    PoolName string
 
-    // Results lists all allocated driver resources.
-    Results []DriverAllocationResult
+    // DeviceName references one device instance via its name in the driver's
+    // resource pool. It must be a DNS label.
+    DeviceName string
 }
 
-// DriverAllocationResult contains vendor parameters and the allocation result for
-// one request.
-type DriverAllocationResult struct {
-    // VendorRequestParameters are the per-request configuration parameters
-    // from the time that the claim was allocated.
-    VendorRequestParameters runtime.Object
+const PoolNameMaxLength = validation.DNS1123SubdomainMaxLength // Same as for a single node name.
+
+// DeviceConfiguration is one entry in a list of configuration pieces for a device.
+type DeviceConfiguration struct {
+    // Admins is true if the source of the piece was a class and thus
+    // not something that a normal user would have been able to set.
+    Admin bool
 
-    AllocationResultModel
+    DriverConfigurationParameters
```

Comment:
We agreed that "configuration" is a one-of, with one option right now the opaque vendor configuration. Standardized alternatives could get added later.

Reply:
I guess I'm still confused as to what was decided for when we embed a struct vs. when we just put the one-of parameters inline. My understanding was that we decided to *always* put the one-of parameters inline...

Reply:
The `ConfigurationParameters` is a one-of struct because it gets embedded in a slice where we need a struct. I'm using `DriverConfigurationParameters` as a struct because it mirrors that other struct.

Inlining a single field also has another challenge. It has to be required, but at the same time has to be a pointer because it might become optional:
```
type DeviceConfiguration struct {
   Admin bool
   
    // Currently required, might become optional when alternatives are added. Check for ni!
    Opaque *runtime.RawExtension
}
```

I find the struct easier to understand.